strawberryfields/engine.py

```diff
@@ -270,8 +270,9 @@ def _broadcast_nones(val, dim):
                     p.reg_refs[k].val = v
 
             # if the program hasn't been compiled for this backend, do it now
-            if p.target != self.backend_name:
-                p = p.compile(self.backend_name, **compile_options) # TODO: shots might be relevant for compilation?
+            target = self.backend.circuit_spec
+            if target is not None and p.target != target:
```

Comment:
is it possible that `p.target` could be set to `target`, but `p` has not actually (yet) been compiled for that target? I'm thinking if someone has blackbird code loaded from an `xbb` file with `target=gaussian` (or whatever), that information will propagate to `p.target`. If the engine is attached to Gaussian backend, then the above line will effectively skip compilation. 

In this case, is it possible that the `.xbb` file will contain incompatible ops (e.g., fock state preparation), but it never gets validated?

Reply:
Looks like compilation is carried out during `sf.io.load()`, so that is ok.

But we could still have that a user does the following:
```
prog = sf.Program(1)
with prog.context as q:
    SomeUnsupportedOp() | q

eng = sf.Engine("gaussian")
prog.target = "gaussian"
eng.run(prog)
```

Reply:
but I think that any user that knew enough to switch `prog.target` like this should not be confused if they receive an error of some kind from `eng.run()` (just not a compilation error)